### PR TITLE
Mesa: reword expression

### PR DESCRIPTION
**Pull request by AndersonTorres**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by K900 — APPROVED** [review]
Overall LGTM but will conflict with #286041

**Comment by AndersonTorres** [comment]
> Overall LGTM but will conflict with #286041

I removed that specific modification. Hope it does not conflict.

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -168,61 +194,62 @@ self = stdenv.mkDerivation {
...
+    (lib.mesonOption "gbm-backends-path" "${libglvnd.driverLink}/lib/gbm:${placeholder "out"}/lib/gbm")
 
     # meson auto_features enables these features, but we do not want them
-    "-Dandroid-libbacktrace=disabled"
+    (lib.mesonEnable "android-libbacktrace" false)
 
-  ] ++ lib.optionals stdenv.isLinux [
-    "-Dglvnd=true"
+    (lib.mesonEnable "libunwind" (!withLibunwind))
+    (lib.mesonEnable "valgrind" (!withValgrind))
```
Comment (vcunat) [inline pkgs/development/libraries/mesa/default.nix:223]: Looks like reversed logic here.

Comment (AndersonTorres) [reply pkgs/development/libraries/mesa/default.nix:223]: Oh, indeed!

**Comment by AndersonTorres** [comment]
@ofborg build

**Comment by AndersonTorres** [comment]
@ofborg eval

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -39,7 +71,6 @@
     [
       "amd" # AMD (aka RADV)
       "microsoft-experimental" # WSL virtualized GPU (aka DZN/Dozen)
-      "nouveau-experimental" # Nouveau (aka NVK)
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:42]: And this

File: pkgs/development/libraries/mesa/default.nix [hunk 3]
```diff
@@ -60,29 +91,29 @@
...
+/** Packaging design:
+  - The basic mesa ($out) contains headers and libraries (GLU is in libGLU now).
+    This or the mesa attribute (which also contains GLU) are small (~ 2 MB, mostly headers)
+    and are designed to be the buildInput of other packages.
+  - DRI drivers are compiled into $drivers output, which is much bigger and
+    depends on LLVM. These should be searched at runtime in
+    "/run/opengl-driver{,-32}/lib/*" and so are kind-of impure (given by NixOS).
+    (I suppose on non-NixOS one would create the appropriate symlinks from there.)
+  - libOSMesa is in $osmesa (~4 MB)
+*/
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:112]: And this

**Comment by K900** [comment]
...and a lot of other things actually.

**Comment by vcunat** [comment]
Yeah well, typically my biggest issue with reformats are these complications.  Harder archeology (e.g. blaming), breakage of other work that hasn't been merged yet, backports, etc.  So I often have a dilemma whether the (arguably) better formatting is worth that.

File: pkgs/development/libraries/mesa/default.nix [hunk 4]
```diff
@@ -1,16 +1,48 @@
...
+, wayland-protocols
+, wayland-scanner
+, xorg
+, zstd
+
 , galliumDrivers ?
   if stdenv.isLinux then
     [
       "d3d12" # WSL emulated GPU (aka Dozen)
-      "kmsro" # special "render only" driver for GPUs without a display controller
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/development/libraries/mesa/default.nix:13]: Fixed.

**Comment by K900** [comment]
I wonder if we can get this to be rebuild-0. That would make merging it much easier.

**Comment by AndersonTorres** [comment]
> Yeah well, typically my biggest issue with reformats are these complications.

Well, I usually cite my own work on rewording Emacs expression: it is hard, complicated and annoying, but someone somehow will need to do it someday.

> sometimes bugs sneak into the rewrite

Well, I usually cite my own work on rewording Emacs expression: it undercovered a sneaky bug where an attribute was set to `null` even when it was required.

>  So I often have a dilemma whether the (arguably) better formatting is worth that.

Well, I usually cite my own work on rewording Emacs expression: it pays on the long run.

> I wonder if we can get this to be rebuild-0. That would make merging it much easier.

The `lib.meson*` functions do some type checking. Maybe it will trigger rebuilds since they are not equivalent to a raw `-Doption=value`.

**Comment by K900** [comment]
It shouldn't, they evaluate to the same thing in the end (or at least they should, I think?).

**Comment by AndersonTorres** [comment]
I understand very few about the Nix evaluator and the dark magic around the Nix expressions from Nixpkgs codebase.
Here is the code for `mesonOption`:

```nix
mesonOption = feature: value:
  assert (lib.isString feature);
  assert (lib.isString value);
  "-D${feature}=${value}";
```

Maybe the assertions are evaluated even when they change nothing in the end.

**Comment by K900** [comment]
They're not going to affect the final .drv file.

**Comment by AndersonTorres** [comment]
Newbie question: how can I verify what is introducing differences between the drv files?

**Comment by K900** [comment]
`nix-diff`

**Comment by AndersonTorres** [comment]
I will do this one step at time. The reword of meson flags looks more important now.

**Comment by K900** [comment]
Doesn't seem to have gotten rid of the rebuild count :(

**Comment by vcunat** [comment]
At a quick glance I see swapped order between `eglPlatforms` and `galliumDrivers`.  (but there might be more differences)

**Comment by AndersonTorres** [comment]
Ah, swapping order again...

**Comment by AndersonTorres** [comment]
Two questions:

1. Does `lib.optional F C` generate a different code from `lib.optional F [ C ]`;

2. Is this line expected?

```nix
++ lib.optional (vulkanLayers != []) "-D vulkan-layers=${builtins.concatStringsSep "," vulkanLayers}";
```

It has a whitespace between `-D` and the Meson flag. 

It will certainly be different from

```
 ++ lib.optional (vulkanLayers != []) (lib.mesonOption "vulkan-layers" (builtins.concatStringsSep "," vulkanLayers))
```

**Comment by K900** [comment]
1. Yes, unless you mean `lib.optionals F [ C ]` in which case no.
2. Meson accepts both ways, if this is what's causing the rebuilds, maybe we can split the change into one part that doesn't cause rebuilds and another part that actually does? Because most changes here should not need to be a rebuild, and it'll be easier to review the interesting parts if they're separated out.

**Comment by AndersonTorres** [comment]
Well, this is now a 66-line diff. Let's see...

**Comment by AndersonTorres** [comment]
YAY! Zero rebuilds!

**Pull request merged**